Spigot-Server-Patches/0167-Anti-Xray.patch

```diff
@@ -0,0 +1,997 @@
+index d59fca4..1da7390 100644
+--- a/src/main/java/net/minecraft/server/NetworkManager.java
++++ b/src/main/java/net/minecraft/server/NetworkManager.java
+@@ -138,8 +138,8 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
+     }
+ 
+     public void sendPacket(Packet<?> packet) {
+-        if (this.isConnected()) {
+-            this.m();
++        if (this.isConnected() && this.m() && !(packet instanceof PacketPlayOutMapChunk && !((PacketPlayOutMapChunk) packet).isReady())) { // Paper - Async-Anti-Xray - Add not ready chunk packets and all packets if the queue contains a not ready chunk packet to the queue to keep the packet-order
++            //this.m(); // Paper - Async-Anti-Xray - Move to if-statement
+             this.a(packet, (GenericFutureListener[]) null);
+         } else {
+             this.j.writeLock().lock();
+@@ -154,8 +154,8 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
+     }
+ 
+     public void sendPacket(Packet<?> packet, GenericFutureListener<? extends Future<? super Void>> genericfuturelistener, GenericFutureListener<? extends Future<? super Void>>... agenericfuturelistener) {
+-        if (this.isConnected()) {
+-            this.m();
++        if (this.isConnected() && this.m() && !(packet instanceof PacketPlayOutMapChunk && !((PacketPlayOutMapChunk) packet).isReady())) { // Paper - Async-Anti-Xray - Add not ready chunk packets and all packets if the queue contains a not ready chunk packet to the queue to keep the packet-order
++            //this.m(); // Paper - Async-Anti-Xray - Move to if-statement
+             this.a(packet, (GenericFutureListener[]) ArrayUtils.add(agenericfuturelistener, 0, genericfuturelistener));
+         } else {
+             this.j.writeLock().lock();
+@@ -210,21 +210,28 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
+ 
+     }
+ 
+-    private void m() {
++    private boolean m() { // Paper - Async-Anti-Xray - void -> boolean used in sendPacket
+         if (this.channel != null && this.channel.isOpen()) {
+             this.j.readLock().lock();
+ 
+             try {
+                 while (!this.i.isEmpty()) {
+-                    NetworkManager.QueuedPacket networkmanager_queuedpacket = (NetworkManager.QueuedPacket) this.i.poll();
++                    NetworkManager.QueuedPacket networkmanager_queuedpacket = (NetworkManager.QueuedPacket) this.i.peek(); // Paper - Async-Anti-Xray - poll() -> peek()
+ 
+-                    this.a(networkmanager_queuedpacket.a, networkmanager_queuedpacket.b);
++                    if (networkmanager_queuedpacket.a instanceof PacketPlayOutMapChunk && !((PacketPlayOutMapChunk) networkmanager_queuedpacket.a).isReady()) { // Paper - Async-Anti-Xray - Return false if the queue contains a not ready chunk packet
++                        return false; // Paper - Aync-Anti-Xray
```

Comment:
It's not that easy as you might think to just queue only the chunk packet and send it later. The order of the packets is sometimes important (eg chunk packet then block updates). That's the reason why I don't call sendPacket(packetPlayOutMapChunk) later or asyncronously when the packet is finished. It is still called synchronous (with the not finished packet), so that the server "knows" that it can now send blockupdates for example. That's reason why all packets are queued in the right order.